providers/implementations/macs/hmac_prov.c

```diff
@@ -59,6 +64,15 @@ struct hmac_data_st {
     int tls_header_set;
     unsigned char tls_mac_out[EVP_MAX_MD_SIZE];
     size_t tls_mac_out_size;
+#ifdef FIPS_MODULE
```

Comment:
I have a question:
is it OK to change structure size when code gets compiled with FIPS *disabled*?
I'm just worried about binary compatibility. The offset of members further below will also change.
but it does not matter perhaps.

also hmac_dup() seems to forget to copy `internal` member from src to dst.

Reply:
> is it OK to change structure size when code gets compiled with FIPS disabled?

This makes no difference. This code is compiled *twice* when configured with "enable-fips". It is compiled once with `FIPS_MODULE` not defined and the resulting code is incorporated into the default provider. It is compiled a second time with `FIPS_MODULE` defined and that code is incorporated into the fips provider. This second version of the code is entirely contained within the fips provider. There are no binary compatibility issues with this.

Reply:
Reasonable question though..

Reply:
> also hmac_dup() seems to forget to copy internal member from src to dst.

This should be covered by this line:

```` c
    *dst = *src;
````

Reply:
OK I see it now. but still would prefer a consistency between kmac and hmac (if possible).
but this is just my preference which I don't insist on.

Reply:
Having the structures change between the two compilations isn't a binary compatibility issue as @mattcaswell noted.  It causes annoyance debugging but provides, what I think is, the significant benefit of ensuring that the FIPS provider/libcrypto boundary remains pristine.  If a structure accidentally crosses this boundary, things start failing.

The annoyance is worst for the random bit generation where both sides of the boundary need to be poked with simultaneously to figure out what's going on.  For other algorithms, it's just one side which is generally okay.